### DISCUSSION TITLE
Vision and Reality, Which?

### DISCUSSION THREAD

**Discussion by bart3nder** [XTLS/Xray-core#2166] [Q&A] [answered]
Hi! Just a quick question:
With the existence of Reality, Is there any reason to use Vision? I'm already using both on my config file (Reality TCP RPRX Vision Flow) and I was wondering If there is no Performance, Security or ... priority to use Vision, I disable it and only use Reality.

**Comment by FranzKafkaYu** [comment, marked as answer]
Reality can eliminate server tls fingerprint characteristics.
xtls-rprx-vision can eliminate  tls-in-tls characteristics.
utls can eliminate client tls fingerprint characteristics.

so if you need the best secure and resistant combination,Reality+xtls-rprx-vision+utls should be the best Choice.

but  this combination can't pass through CDN，it's a drawback here.

**Reply by bart3nder** [reply]
Great explanation, Thanks!

**Reply by houmie** [reply]
Hello @FranzKafkaYu 

1) Is it UTLS in this combination (Reality+xtls-rprx-vision+utls) that makes it not able to pass though CDN such as CloudFlare? Which combos are allowed? Is there a tutorial for how to use a CDN for that?

2) Is this Caddy plugin able to retrieve Cloudflare CDN IPs? https://github.com/WeidiDeng/caddy-cloudflare-ip
So if I connect though a Xray/Caddy VPN with this plugin enabled, I would get a Cloudflare IPs assigned instead of my VPS IP address?  (So under https://ipleak.net which Ip address would I see?)

Thank you so much,

**Reply by jiushibushuo** [reply]
> Hello @FranzKafkaYu
> 
> 1. Is it UTLS in this combination (Reality+xtls-rprx-vision+utls) that makes it not able to pass though CDN such as CloudFlare? Which combos are allowed? Is there a tutorial for how to use a CDN for that?
> 2. Is this Caddy plugin able to retrieve Cloudflare CDN IPs? https://github.com/WeidiDeng/caddy-cloudflare-ip
>    So if I connect though a Xray/Caddy VPN with this plugin enabled, I would get a Cloudflare IPs assigned instead of my VPS IP address?  (So under https://ipleak.net which Ip address would I see?)
> 
> Thank you so much,

any of the 3 components stops you from using CDN, if you are obsessed with CDN, try ws+tls then

**Reply by houmie** [reply]
@jiushibushuo Thank you for your reply.  One last question about fingerprint:

[Here](https://cscot.pages.dev/2023/03/02/Xray-REALITY-tutorial/#Xray-core-command-line-client-for-Windows) I saw that it says:
```fingerprint: Use uTLS library to emulate client TLS fingerprint.```

Do you know if I need the uTLS library installed on my iPhone or Android for enabling UTLS? Or do the Xray clients automatically include this?

**Reply by 2moravin** [reply]
> Reality can eliminate server tls fingerprint characteristics. xtls-rprx-vision can eliminate tls-in-tls characteristics. utls can eliminate client tls fingerprint characteristics.
> 
> so if you need the best secure and resistant combination,Reality+xtls-rprx-vision+utls should be the best Choice.
> 
> but this combination can't pass through CDN，it's a drawback here.

How can Reality+xtls-rprx-vision+utls be used simultaneously?

**Comment by ghost** [comment]
> Reality can eliminate server tls fingerprint characteristics.
> xtls-rprx-vision can eliminate  tls-in-tls characteristics.
> utls can eliminate client tls fingerprint characteristics.
> 
> so if you need the best secure and resistant combination,Reality+xtls-rprx-vision+utls should be the best Choice.
> 
> but  this combination can't pass through CDN，it's a drawback here.

@FranzKafkaYu 

I wonder if XTLS_RPRX_Vision flow can only be applied to TCP transport or if it is compatible with H2 and gRPC?

**Reply by FranzKafkaYu** [reply]
> > Reality can eliminate server tls fingerprint characteristics.
> > xtls-rprx-vision can eliminate  tls-in-tls characteristics.
> > utls can eliminate client tls fingerprint characteristics.
> > so if you need the best secure and resistant combination,Reality+xtls-rprx-vision+utls should be the best Choice.
> > but  this combination can't pass through CDN，it's a drawback here.
> 
> @FranzKafkaYu
> 
> I wonder if XTLS_RPRX_Vision flow can only be applied to TCP transport or if it is compatible with H2 and gRPC?

tcp only.
